### ISSUE TITLE
don't register pandas mpl unit converters upon import

### ISSUE THREAD

**Issue by changhiskhan** [pandas-dev/pandas#2579]
register it upon plotting?

http://stackoverflow.com/questions/13988111/importing-pandas-in-python-changes-how-matplotlib-handles-datetime-objects

**Comment by TomAugspurger** [comment]
We should be able to wrap something like @ruidc's code up in an option. I'll see if I have time today.

**Comment by ocehugo** [comment]
@TomAugspurger,

I just want do add more superannoying this is and urge the fix so we can be all happy fellas. 

All packages that import pandas cause this.

So this breaks all calls with datetime objects in matplotlib with old dates.

In the case one might think: but why you don't go with the flow and use pandas to deal with dates!? Is much better... OK ... but wait I can't! Since pandas do not support date_ranges before 1978 (out of bounds).

```python
import pandas as pd
xtime = pd.date_range('1/1/1976','11/1/1976',freq='M')
OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 1676-01-01 00:00:00
```

Using the kludge above works but still a kludge hidden in here.

**Comment by jreback** [comment]
what version of pandas are you showing above ?

**Comment by ocehugo** [comment]
see #13713

**Comment by jreback** [comment]
your example of using date_range works fine
you have a copy paste error

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/timeseries.html#representing-out-of-bounds-spans

**Comment by TomAugspurger** [comment]
@ocehugo I ran out of time today, so feel free to work on it. I think a pd.option to toggle the regiration is the way to handle it.

**Comment by ocehugo** [comment]
@jreback yeah, pasting error (should be 1676 in xtime). Not a pandas heavy user so didn't remembered the period_range method ( I could do that but this would required me to change all code related so easier to use the kludge). @TomAugspurger, as said not a pandas guy but if you tell me where this should be I could pull something. I believe the kludge above should be default to not conflict with other packages but don't know the consequences of that.

just to stress out , I have some source code that use xarray and others with statsmodels. Just the fact of importing xarray makes all my matplotlib plotting functions not to work.

**Comment by TomAugspurger** [comment]
@ocehugo thanks for taking this.

The goal is to get something like `pd.options.plotting.register_converters` (that name isn't set in stone). That can be set to either True or False.
The option you'll add will be in https://github.com/pydata/pandas/blob/master/pandas/core/config_init.py

For backwards compatibility the default will have to be True (register the converters).
The documentation for the config options is at the top of [this file](https://github.com/pydata/pandas/blob/fc16f1fd21aee163e93e5713a0676f7a79838897/pandas/core/config.py). You'll want to use a callback that calls `converters.register` (see below) each time the option is set / reset.

The actual converters are defined in https://github.com/pydata/pandas/blob/fc16f1fd21aee163e93e5713a0676f7a79838897/pandas/tseries/converter.py#L53

and used in https://github.com/pydata/pandas/blob/fc16f1fd21aee163e93e5713a0676f7a79838897/pandas/tools/plotting.py#L35

I would add an argument to `converters.register`

```
def register(present=True):
    pairs = [
        (lib.Timestamp, DatetimeConverter()),
        (Period, PeriodConverter()),
        ...  # the rest of the converters
    ]
    for key, value in pairs:
        if present:
            units.registry[key] = value
        else:
            units.registry.pop(key, None)
```

And then tests + documentation 😄 Hopefully not too much work, but let me know if you have questions.

**Comment by tacaswell** [comment]
@ocehugo I am impressed by the bread-crumb trail on this issue.

**Comment by jorisvandenbossche** [comment]
An option to configure whether pandas converters are registered for matplotlib or not is certainly a good idea. But can't we also just fix are DatetimeConverter to actually work with all `datetime.datetime` values?  

The following one-line change seems to fix the example from @ocehugo : https://github.com/pydata/pandas/commit/095a2ef6cf696856a79beb1fdd42b2cef89a1a8f (just converting the values itself when `to_datetime` failed) (but maybe I am missing the complexity of the issue)

You still get the adapted axis formatting from pandas (which you could then turn off with the option), but at least the plots would work.

**Comment by ocehugo** [comment]
@tacaswell yeah, messed up with issues in almost all related packages since I was not expecting that importing a package would create such a big issue in almost all my source code ,since some basic packages need to support pandas, they need to import it so the problem was all around and such i went to blame matplotlib first...statsmodels...pandas). Should have investigate further before, but was pretty much present in almost all tests that i did until I had to go to raw jupyter without my default profile).

@TomAugspurger will take a look at that but not until next week

**Comment by jorisvandenbossche** [comment]
@ocehugo If pandas would still register its own converter, but if this would *not* break code that would run without having imported pandas, would that be OK for you? 
There will still be differences (eg in the axis formatting), but I would suppose it would solve the biggest problem?

**Comment by jorisvandenbossche** [comment]
This issue has actually been solved in 0.21.0 (not registering the unit converters by default on pandas import, see https://github.com/pandas-dev/pandas/pull/17710). 
However, giving the many feedback, we will revert this change in 0.21.1, see https://github.com/pandas-dev/pandas/issues/18301. Temporarily, to allow a more graceful deprecation period. 

But as you currently get a deprecation warning (so are notified of using the converters), and this deprecation will eventually be removed (and we have already another issue to track the removal of deprecations) which will remove the automatic registering, I am already closing this issue.
